libcxx/docs/DesignDocs/ABIGuarantees.rst

```diff
@@ -0,0 +1,199 @@
+=======================
+libc++'s ABI Guarantees
+=======================
```

Comment:
I wonder about the location of this file. Since we call them Guarantees I feel this is aimed at our users and our developers.  Reading the document gives me the same vibes; aiming at users and developers. So I would like to have this information to be at the user documentation.

Maybe adding it to https://libcxx.llvm.org/DesignDocs/ABIVersioning.html would be better.

Reply:
I don't think `ABIVersioning` is the right place, since this document isn't really about versioning. (And FWIW I think we need to re-work our ABI versioning story more generally). I'm also not really sure whether we consider our design docs to be user-facing or not. I feel like some are definitely user-facing while others clearly aren't.

Reply:
Our ABI macros are intended to be used by vendors, not by users. I think that is where we should document them and I think we should word that documentation with that in mind.

We should *not* document these ABI macros in any location that makes them look as if users can modify them. I actually see users defining ABI macros they shouldn't define on a regular basis, only to wonder why their code suddenly starts breaking when they trip up an ABI landmine. We could potentially also clarify that those are "internal" macros by changing their name to include something like `INTERNAL`, or `IMPL`, or `VENDOR`.